### PR TITLE
Sink implementation for Google Cloud Storage

### PR DESCRIPTION
**Pull request by CharukaK**
## Purpose
This PR contains the Sink implementation for Google Cloud Storage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Test environment
JDK v1.8.0_211

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,39 @@
 # siddhi-io-gcs
 Extension which can be used to receive/publish event data from/to Google Cloud Storage bucket
+
+**Using the extension as a [java library](https://siddhi-io.github.io/siddhi/documentation/siddhi-5.x/user-guide-5.x/#using-siddhi-as-a-java-library)**
```
Comment (niveathika) [inline README.md]: Use latest README format in extensions

File: component/pom.xml [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Private-Package>
+                            com.google.*,
+                            org.threeten.bp.*,
+                            io.opencensus.*,
+                            io.grpc.*,
+                        </Private-Package>
+                        <Export-Package>
+                            io.siddhi.extension.io.gcs.*,
+                            io.siddhi.extension.io.gcs.*
```
Comment (niveathika) [inline component/pom.xml]: export-package is given twicw

File: component/pom.xml [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                        </Export-Package>
+                        <Import-Package>
+                            com.google.storage.*;version="${gcs.client.version.range}",
+                            io.siddhi.annotation.*;version="${siddhi.version.range}",
+                            io.siddhi.core.*;version="${siddhi.version.range}",
```
Comment (niveathika) [inline component/pom.xml:181]: Import siddhi core once

File: component/src/main/java/io/siddhi/extension/io/gcs/sink/GCSSink.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+import java.nio.ByteBuffer;
+
+/**
+ * GCSSink implementation of Google Cloud Storage extension.
+ */
+@Extension(
+        name = "google-cloud-storage",
+        namespace = "sink",
+        description = " ",
+        parameters = {
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/io/gcs/sink/GCSSink.java:49]: Description

File: component/src/main/java/io/siddhi/extension/io/gcs/sink/GCSSink.java [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+    /**
+     * Returns the list of classes which this sink can consume.
+     * Based on the type of the sink, it may be limited to being able to publish specific type of classes.
+     * For example, a sink of type file can only write objects of type String .
+     *
+     * @return array of supported classes , if extension can support of any types of classes
+     * then return empty array .
+     */
+    @Override
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/io/gcs/sink/GCSSink.java:119]: Remove default comments

File: component/src/main/java/io/siddhi/extension/io/gcs/sink/internal/beans/GCSSinkConfig.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+                    break;
+                case GCSConstants.CONTENT_TYPE:
+                    this.contentType = optionHolder.validateAndGetStaticValue(GCSConstants.CONTENT_TYPE);
+                    break;
+                case GCSConstants.STORAGE_CLASS:
+                    this.storageClass = StorageClass.valueOf(optionHolder
+                            .validateAndGetStaticValue(GCSConstants.STORAGE_CLASS));
+                    break;
+                case GCSConstants.ENABLE_VERSIONING:
+                    this.enableVersioning = Boolean.parseBoolean(
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/io/gcs/sink/internal/beans/GCSSinkConfig.java:58]: default values?

File: component/src/test/resources/testng.xml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
+
+<suite name="Siddhi-IO-GCS-Test-Suite">
+    <test name="Siddhi-io-google cloud storage-tests" enabled="true" preserve-order="true" parallel="false">
+        <classes>
+<!--            <class name="io.siddhi.extension.io.gcs.sink.TestCaseOfGCSSink"/>-->
```
Comment (niveathika) [inline component/src/test/resources/testng.xml:10]: Add a comment about disabling tests

File: docs/images/siddhi-logo-w.svg [hunk 8]
```diff
@@ -0,0 +1,163 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<!--
```
Comment (niveathika) [inline docs/images/siddhi-logo-w.svg:2]: Use latest artefacts for docs, add a license for docs/license

File: mkdocs.yml [hunk 9]
```diff
@@ -0,0 +1,30 @@
+site_name: Siddhi IO GCS
+site_description: Siddhi IO GCS Extension
```
Comment (niveathika) [inline mkdocs.yml:2]: Complete as same as the latest extensions

File: pom.xml [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+            <dependency>
+                <groupId>io.siddhi.extension.map.json</groupId>
+                <artifactId>siddhi-map-json</artifactId>
+                <version>${siddhi.map.json.version}</version>
+                <scope>test</scope>
+            </dependency>
+            <dependency>
+                <groupId>org.testng</groupId>
+                <artifactId>testng</artifactId>
+                <version>6.11</version>
```
Comment (niveathika) [inline pom.xml]: Move versions to properties

File: component/src/test/java/io/siddhi/extension/io/gcs/sink/TestCaseOfGCSSink.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
+package io.siddhi.extension.io.gcs.sink;
```
Comment (niveathika) [inline component/src/test/java/io/siddhi/extension/io/gcs/sink/TestCaseOfGCSSink.java:20]: License

File: component/src/main/java/io/siddhi/extension/io/gcs/util/ServiceClient.java [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+                        .setCredentials(GoogleCredentials
+                                .fromStream(new FileInputStream(
+                                        new File(config.getCredentialFilePath())))).build().getService();
+            }
+        } catch (StorageException e) {
+            throw new ConnectionUnavailableException(e.getMessage(), e);
+        } catch (IOException e) {
+            logger.error(e.getMessage(), e);
+        }
+
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/io/gcs/util/ServiceClient.java]: remove unnecessary lines

**Review by niveathika — CHANGES_REQUESTED** [review]

**Review by niveathika — APPROVED** [review]

File: .gitignore [hunk 13]
```diff
@@ -1,3 +1,261 @@
...
+# When using Gradle or Maven with auto-import, you should exclude module files,
+# since they will be recreated, and may cause churn.  Uncomment if using
+# auto-import.
+# .idea/modules.xml
+# .idea/*.iml
+# .idea/modules
+# *.iml
+# *.ipr
+
+# CMake
```
Comment (niruhan) [inline .gitignore:234]: shall we remove these placeholders?

File: checkstyle-suppressions.xml [hunk 14]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (niruhan) [inline checkstyle-suppressions.xml:1]: can we safely remove this file?

File: component/pom.xml [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+        </dependency>
+        <dependency>
+            <groupId>io.opencensus</groupId>
+            <artifactId>opencensus-contrib-http-util</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.grpc</groupId>
+            <artifactId>grpc-context</artifactId>
+        </dependency>
+
```
Comment (niruhan) [inline component/pom.xml:89]: shall we remove the newline

File: component/src/main/java/io/siddhi/extension/io/gcs/sink/GCSSink.java [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+                        name = GCSConstants.BUCKET_ACL,
+                        type = DataType.STRING,
+                        optional = true,
+                        defaultValue = "EMPTY_STRING",
+                        description = "Access Control List for the bucket level ACL defined as a key value pair list" +
+                                " defined as \"'<key>:<value>','<key>:<value>'\""
+                )
+        },
+        examples = {
+                @Example(
```
Comment (niruhan) [inline component/src/main/java/io/siddhi/extension/io/gcs/sink/GCSSink.java:96]: shall we add one example with optional parameters?

File: component/src/main/java/io/siddhi/extension/io/gcs/sink/GCSSink.java [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+    @Override
+    public void connect() throws ConnectionUnavailableException {
+        config.setMapType(getMapper().getType());
+        publisher.initializeServiceClient();
+    }
+
+    /**
+     * Called after all publishing is done, or when a ConnectionUnavailableException is thrown.
+     */
+    @Override
```
Comment (niruhan) [inline component/src/main/java/io/siddhi/extension/io/gcs/sink/GCSSink.java:183]: can we remove this overriding?

File: component/src/main/java/io/siddhi/extension/io/gcs/util/ServiceClient.java [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+                        break;
+                    case "writer":
+                        createAclForUser(item.getKey(), Acl.Role.WRITER);
+                        break;
+                    default:
+                        // not a valid type of Permission.
+                }
+            }
+        }
+
```
Comment (dilini-muthumala) [inline component/src/main/java/io/siddhi/extension/io/gcs/util/ServiceClient.java:110]: It's better to remove unnecessary empty lines.

File: component/src/main/java/io/siddhi/extension/io/gcs/util/ServiceClient.java [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+        return this;
+    }
+
+    public void uploadObject(String objectName, String objectContent) {
+        BlobId blobId = BlobId.of(config.getBucketName(), objectName);
+        BlobInfo blobInfo = BlobInfo.newBuilder(blobId)
+                .setContentType(getContentType(((GCSSinkConfig) config).getMapType()))
+                .build();
+
```
Comment (dilini-muthumala) [inline component/src/main/java/io/siddhi/extension/io/gcs/util/ServiceClient.java:119]: It's better to remove unnecessary empty lines.

**Pull request merged**